src/java.base/share/classes/java/lang/Module.java

```diff
@@ -256,25 +258,74 @@ public ModuleLayer getLayer() {
     /**
      * Update this module to allow access to restricted methods.
      */
-    Module implAddEnableNativeAccess() {
+    synchronized Module implAddEnableNativeAccess() {
         enableNativeAccess = true;
         return this;
     }
 
     /**
-     * Update all unnamed modules to allow access to restricted methods.
+     * Returns {@code true} if this module can access
+     * <a href="foreign/package-summary.html#restricted"><em>restricted</em></a> methods.
+     *
+     * @return {@code true} if this module can access <em>restricted</em> methods.
      */
-    static void implAddEnableNativeAccessAllUnnamed() {
-        ALL_UNNAMED_MODULE.enableNativeAccess = true;
+    @PreviewFeature(feature=PreviewFeature.Feature.FOREIGN)
+    public boolean isNativeAccessEnabled() {
+        Module target = enableNativeAccessHolder();
+        synchronized(target) {
+            return target.enableNativeAccess;
+        }
```

Comment:
enableNativeAccess is `@Stable` and I'm trying to understand the access. ensureNativeAccess will only synchronize when false, isNativeAccessEnabled synchronizes always. Is that intentional?

Reply:
That is intentional. `ensureNativeAccess` can be called in hot paths, so it's crucial we only synchronized when needed (so we do an optimistic plain read and lock if we see `false`). The new API predicate on `Module` is a public API, so we do not have the same performance constraints there, and we can make the implementation more intuitive.